File: dxbase/src/main/java/com/dxmovie/dxbase/binding/viewadapter/viewpager/ViewAdapter.java
package com.dxmovie.dxbase.binding.viewadapter.viewpager;



import androidx.databinding.BindingAdapter;
import androidx.viewpager.widget.ViewPager;

import com.dxmovie.dxbase.binding.command.BindingCommand;
import com.dxmovie.dxbase.binding.command.BindingCommandWithParams;

/**
 */
public class ViewAdapter {
    @BindingAdapter(value = {"onPageScrolledCommand", "onPageSelectedCommand", "onPageScrollStateChangedCommand"}, requireAll = false)
    public static void onScrollChangeCommand(final ViewPager viewPager,
                                             final BindingCommandWithParams<ViewPagerDataWrapper> onPageScrolledCommand,
                                             final BindingCommandWithParams<Integer> onPageSelectedCommand,
                                             final BindingCommandWithParams<Integer> onPageScrollStateChangedCommand) {
        viewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            private int state;

            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
                if (onPageScrolledCommand != null) {
                    onPageScrolledCommand.executeAction(new ViewPagerDataWrapper(position, positionOffset, positionOffsetPixels, state));
                }
            }

            @Override
            public void onPageSelected(int position) {
                if (onPageSelectedCommand != null) {
                    onPageSelectedCommand.executeAction(position);
                }
            }

            @Override
            public void onPageScrollStateChanged(int state) {
                this.state = state;
                if (onPageScrollStateChangedCommand != null) {
                    onPageScrollStateChangedCommand.executeAction(state);
                }
            }
        });

    }

    public static class ViewPagerDataWrapper {
        public float positionOffset;
        public float position;
        public int positionOffsetPixels;
        public int state;

        public ViewPagerDataWrapper(float position, float positionOffset, int positionOffsetPixels, int state) {
            this.positionOffset = positionOffset;
            this.position = position;
            this.positionOffsetPixels = positionOffsetPixels;
            this.state = state;
        }
    }
}

File: dxbase/src/main/java/com/dxmovie/dxbase/binding/viewadapter/viewpager/ViewAdapter.kt
package com.dxmovie.dxbase.binding.viewadapter.viewpager

import androidx.databinding.BindingAdapter
import androidx.viewpager.widget.ViewPager
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import com.dxmovie.dxbase.binding.command.BindingCommandWithParams

class ViewAdapter {

    companion object{
        @BindingAdapter(value = ["onPageScrolledCommand", "onPageSelectedCommand", "onPageScrollStateChangedCommand"], requireAll = false)
        fun onScrollChangeCommand(viewPager: ViewPager,
                                  onPageScrolledCommand: BindingCommandWithParams<ViewPagerDataWrapper?>?,
                                  onPageSelectedCommand: BindingCommandWithParams<Int?>?,
                                  onPageScrollStateChangedCommand: BindingCommandWithParams<Int?>?) {
            viewPager.addOnPageChangeListener(object : OnPageChangeListener {
                private var state = 0
                override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
                    onPageScrolledCommand?.executeAction(ViewPagerDataWrapper(position.toFloat(), positionOffset, positionOffsetPixels, state))
                }

                override fun onPageSelected(position: Int) {
                    onPageSelectedCommand?.executeAction(position)
                }

                override fun onPageScrollStateChanged(state: Int) {
                    this.state = state
                    onPageScrollStateChangedCommand?.executeAction(state)
                }
            })
        }
    }

    class ViewPagerDataWrapper(var position: Float, var positionOffset: Float, var positionOffsetPixels: Int, var state: Int)
}